### PR TITLE
bpo-30529: Fix errors for invalid whitespaces in f-string subexpressions.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
'invalid character in identifier' now is raised instead of
'f-string: empty expression not allowed' if a subexpression contains
only whitespaces and they are not accepted by Python parser.

### CODE REVIEW

File: Python/ast.c [hunk 1]
```diff
@@ -4274,49 +4274,32 @@ fstring_compile_expr(const char *expr_start, const char *expr_end,
...
-        return NULL;
-    len = PyUnicode_GET_LENGTH(o);
-    kind = PyUnicode_KIND(o);
-    data = PyUnicode_DATA(o);
-    for (i = 0; i < len; i++) {
-        if (!Py_UNICODE_ISSPACE(PyUnicode_READ(kind, data, i))) {
-            all_whitespace = 0;
+    for (s = expr_start; s != expr_end; s++) {
+        char c = *s;
+        if (!(c == ' ' || c == '\t' || c == '\n' || c == '\014')) {
```
Comment (ericvsmith) [inline Python/ast.c]: Wouldn't '\014' (formfeed) be better written as '\f'?

File: Lib/test/test_fstring.py [hunk 2]
```diff
@@ -279,6 +279,7 @@ def test_missing_expression(self):
                              "f'{ !r}'",
                              "f'{10:{ }}'",
                              "f' { } '",
+                             "f'''{\t\x0c\r\n}'''",
```
Comment (ericvsmith) [inline Lib/test/test_fstring.py]: \x0c is \f, isn't it? I'd use that, instead.

File: Lib/test/test_fstring.py [hunk 3]
```diff
@@ -300,6 +301,9 @@ def test_missing_expression(self):
                              "f'{:x'",
                              ])
 
+        self.assertAllRaise(SyntaxError, 'invalid character in identifier',
```
Comment (ericvsmith) [inline Lib/test/test_fstring.py]: Can you add a comment here that says that eval("\xa0") raises SyntaxError with the "invalid character in identifier" text? I think that it's important to know that what you're doing here is trying to match the Python parser. (I realize that, with your change, you're "matching" by actually using the Python parser.)

Or, maybe also add 'eval("\xa0")' to the array to make it obvious that we want the same error raised in both cases?

File: Python/ast.c [hunk 4]
```diff
@@ -4274,49 +4274,32 @@ fstring_compile_expr(const char *expr_start, const char *expr_end,
...
-    o = PyUnicode_DecodeUTF8(expr_start, expr_end-expr_start, NULL);
-    if (o == NULL)
-        return NULL;
-    len = PyUnicode_GET_LENGTH(o);
-    kind = PyUnicode_KIND(o);
-    data = PyUnicode_DATA(o);
-    for (i = 0; i < len; i++) {
-        if (!Py_UNICODE_ISSPACE(PyUnicode_READ(kind, data, i))) {
-            all_whitespace = 0;
+    for (s = expr_start; s != expr_end; s++) {
```
Comment (ericvsmith) [inline Python/ast.c:4290]: Can you add a comment here that you're skipping the same whitespace that the Python parser recognizes, and that's why we're specifically not calling Py_UNICODE_ISSPACE? I can see someone (might be me!) and not realizing that Python itself doesn't use Py_UNICODE_ISSPACE and trying to add it back.

**Review by ericvsmith — CHANGES_REQUESTED** [review]
In general, it looks good, although I think it could be improved with the few tweaks I've mentioned. Not needing to convert from UTF-8 is definitely an improvement. Thanks!

File: Lib/test/test_fstring.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_fstring.py]: Yes. For unknown reasons the repr of `'\f'` is `"'\x0c'"`.

File: Python/ast.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Python/ast.c]: The Python parser uses `'\014'`. I prefer writing it as `'\f'` too.

File: Lib/test/test_fstring.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/test/test_fstring.py]: Done.

File: Python/ast.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Python/ast.c:4290]: Done.

**Review by ericvsmith — APPROVED** [review]

**Pull request merged**